lib/api/readable.js

```diff
@@ -267,14 +265,35 @@ function consumeStart (consume) {
   }
 }
 
+/**
+ * @param {Buffer[]} chunks
+ * @param {number} length
+ */
+function chunksDecode (chunks, length) {
+  if (chunks.length === 0 || length === 0) {
+    return ''
+  }
+  const buffer = chunks.length === 1 ? chunks[0] : Buffer.concat(chunks, length)
+
+  const start =
+   buffer.length >= 3 &&
+   // Skip BOM.
+      buffer[0] === 0xef &&
+      buffer[1] === 0xbb &&
+      buffer[2] === 0xbf
+     ? 3
+     : 0
+  return buffer.utf8Slice(start, buffer.length - start)
```

Comment:
@KhafraDev Do we need to do a `toWellFormed` here?

Reply:
@ronag It looks like no problem.
```js
import { request, MockAgent, setGlobalDispatcher } from "./index.js";

const Q = "\ud800\uda04"

const agent = new MockAgent();

setGlobalDispatcher(agent);

const mockPool = agent.get("http://localhost:3000");
mockPool.intercept({ path: "/foo" }).reply(200, Q);

const requestResponse = await (
  await request("http://localhost:3000/foo")
).body.text();

console.log(requestResponse === Q); // false
console.log(requestResponse === Q.toWellFormed()); // true
```

Reply:
@ronag It looks like no problem.
```js
import { request, MockAgent, setGlobalDispatcher, Response } from "./index.js";
const Q = "\ud800\uda04";
const agent = new MockAgent();
setGlobalDispatcher(agent);
const mockPool = agent.get("http://localhost:3000");
mockPool.intercept({ path: "/foo" }).reply(200, Q);
const requestResponse = await (
  await request("http://localhost:3000/foo")
).body.text();
console.log(requestResponse === Q); // false
console.log(requestResponse === Q.toWellFormed()); // true
console.log(requestResponse === await new Response(Q).text()); // true
console.log(requestResponse === new TextDecoder().decode(new TextEncoder().encode(Q))) // true
console.log(requestResponse === Buffer.from(Q).toString()) // true
```

Reply:
we do

Reply:
@KhafraDev Why is it needed? Can you give an example?

Reply:
maybe utf8slice does the same, idk

Reply:
`toWellFormed`:
>The toWellFormed() method of [String](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String) values returns a string where all [lone surrogates](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String#utf-16_characters_unicode_code_points_and_grapheme_clusters) of this string are replaced with the Unicode replacement character U+FFFD.

The difference seems to be that `toUSVString` removes all surrogates, meanwhile `toWellFormed` is just the `lone-surrogates`. I'd imagine we can choose based on the use case we are covering.

Reply:
so we should either stop calling these spec compliant body mixins or remove the utf8slice because the 2 aren't compatible?

Reply:
@metcoder95 The input is UTF-8 which does not have a concept of surrogate. The closest you can come to is an invalid UTF-8 byte sequence that would translate into a lone surrogate if you just did UTF-8 to UTF-16 transcoding in a broken manner (e.g., using the non-standard WTF-8 interpretation) but `utf8Slice` does not accept invalid UTF-8, it uses replacement characters instead... So it cannot produce something that would lead to a lone surrogate.

```
> Buffer.from(Buffer.from([0xed,0xb8,0x80]).utf8Slice(0,3).toWellFormed())
<Buffer ef bf bd ef bf bd ef bf bd>
> Buffer.from(Buffer.from([0xed,0xb8,0x80]).utf8Slice(0,3))
<Buffer ef bf bd ef bf bd ef bf bd>
> Buffer.from(Buffer.from([0xed,0xb8,0x80]).toString())
<Buffer ef bf bd ef bf bd ef bf bd>
> Buffer.from(util.toUSVString(Buffer.from([0xed,0xb8,0x80])))
<Buffer ef bf bd ef bf bd ef bf bd>
```

Lone surrogates are a feature of UTF-16. The Unicode standard specifically forbids having a lone surrogate in UTF-8:

https://www.unicode.org/faq/utf_bom.html#utf8-5

You may argue that the input might be ill-formed, and indeed, it might be, but the functions we are using (`toString()`, `utf8Slice()`) do transcoding with replacement, so they 'fix' the input.

Here is a surrogate pair:
```
> Buffer.from([0x3d,0xd8, 0x00,0xde]).toString("UTF-16LE")
'😀'
```

Here is a lone surrogate pair:

```
> Buffer.from([0x3d,0xd8]).toString("UTF-16LE")
'\ud83d'
```

But in the code that we are considering here, we seem to assume an UTF-8 input. So I don't understand.

Reply:
@KhafraDev I personally do not know, that's why I am raising the question.